File: src/database/CartaoDAO.java
package database;

import database.core.CoreDAO;
import java.sql.Connection;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

import java.util.ArrayList;
import java.util.List;

import modelo.cliente.Cartao;
import modelo.cliente.Cliente;

public abstract class CartaoDAO extends CoreDAO {

    public static int create(Cartao cartao) throws ClassNotFoundException, SQLException {

        // Obtenha a conexão com o BD;
        Connection conexao = getConnection();

        // Forme a string básica de sql;
        String sql = "INSERT INTO cartao" +
                "(nome_titular, validade, bandeira, numero, tipo)" +
                " VALUES (?, ?, ?, ?, ?)";

        PreparedStatement st = conexao.prepareStatement (sql, Statement.RETURN_GENERATED_KEYS);

        // Defina os valores que ocuparão as '?' na ordem acima;
        st.setString(1, cartao.getTitular());
        st.setDate(2, new Date(cartao.getDataValid().getTime()));
        st.setString(3, cartao.getBandeira());
        st.setLong(4, cartao.getNumero());
        st.setString(5, String.valueOf(cartao.getTipo()));

        // Execute o INSERT e receba o ID do cartão cadastrado no BD;
        st.executeUpdate();
        int id = getIdAtCreate(st);

        st.close();
        conexao.close();

        return id;
    }

    /**
     * Retorna um conjunto de cartões relacionados a um cliente;
     * @param cliente cliente usuário do(s) cartão(ões);
     * @return Conjunto de cartões relacionados ao cliente;
     * @throws java.sql.SQLException
     * @throws java.lang.ClassNotFoundException
     */
    //TODO Consertar
    public static List<Cartao> readCartoesByCliente(Cliente cliente) throws SQLException, ClassNotFoundException {

        List<Cartao> cartoes = new ArrayList<>();

        // Obtenha a conexão com o BD;
        Connection conexao = getConnection();

        // Forme a string sql;
        String sql = "SELECT c.id, c.nome_titular, c.validade, c.bandeira, c.numero," +
                "c.tipo FROM utiliza AS ut INNER JOIN cartao AS c" +
                "ON ut.fk_cartao = c.id WHERE ut.fk_cliente = ?";

        PreparedStatement st = conexao.prepareStatement (sql);
        st.setInt(1, cliente.getId());
        
        ResultSet rs = st.executeQuery();

        // Enquanto houver algum cartão resultado da busca;
        while (rs.next()) {

            int id = rs.getInt("id");
            String bandeira = rs.getString("bandeira");
            Date dataValid = rs.getDate("validade");
            long numero = rs.getLong("numero");
            String titular = rs.getString("nome_titular");
            char tipo = rs.getString("tipo").charAt(0);

            cartoes.add(new Cartao(id, bandeira, dataValid, numero, titular, tipo));
        }

        st.close();
        conexao.close();

        return cartoes;
    }
}

File: extra_python_popular_base/DAO/CartaoDAO.py
import psycopg2

from extra_python_popular_base.DAO.DBCore import DBCore
from extra_python_popular_base.modelo.Cartao import Cartao


class CartaoDAO():

	__nome_tabela = "cartao"
	__cols = ["nome_titular", "validade", "bandeira", "numero", "tipo"]

	@staticmethod
	def get_insert_all_sql(cartoes=[Cartao]):

		cols = CartaoDAO.get_nome_colunas()
		nome_tabela = CartaoDAO.get_nome_tabela()

		sql = "INSERT INTO %s (%s) VALUES" % (nome_tabela, (', '.join(cols)))
		sqls_value = []

		for c in cartoes:
			vals = [c.titular, c.validade, c.bandeira, c.numero, c.tipo]
			sqls_value.append(DBCore.format_sql_value(vals))

		sql = sql + '\n' + (',\n'.join(sqls_value)) + ';'
		return sql

	@staticmethod
	def insert_n(data_base, cartoes=[Cartao]):

		try:
			sql_n_cartoes = CartaoDAO.get_insert_all_sql(cartoes)
			nome_tab = CartaoDAO.get_nome_tabela()

			print("\n", sql_n_cartoes)

			return data_base.insert_all(sql_n_cartoes, nome_tab, "id", 0)

		except psycopg2.ProgrammingError:
			raise

	@staticmethod
	def insert(data_base, cartao=Cartao):

		def get_insert_sql(c=Cartao):
			cols = ["nome_titular", "validade", "bandeira", "numero", "tipo"]
			vals = [c.titular, c.validade, c.bandeira, c.numero, c.tipo]
			sql = DBCore.format_sql_insert("cartao", cols, vals, "id")
			return sql

		cartao.id = data_base.insert(get_insert_sql(cartao))
		return cartao.id

	@staticmethod
	def remove(data_base, cartao_id):

		def get_remove_sql(c_id):
			sql = "DELETE FROM cartao WHERE id = %d;" %c_id
			return sql

		data_base.execute(get_remove_sql(cartao_id))

	@staticmethod
	def get_nome_tabela():
		return CartaoDAO.__nome_tabela

	@staticmethod
	def get_nome_colunas():
		return CartaoDAO.__cols